src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp

```diff
@@ -90,18 +79,105 @@ class SSELoss
     gradients = (input.row(1) - input.row(0)).t();
     hessians = arma::vec(input.n_cols, arma::fill::ones);
 
-    return std::pow(ApplyL1(arma::accu(gradients)), 2) /
-        (arma::accu(hessians) + lambda);
+    // Calculates total sum of gradients and hessians.
+    gTotal = arma::accu(gradients);
+    hTotal = arma::accu(hessians);
+
+    return std::pow(ApplyL1(gTotal), 2) / (hTotal + lambda);
+  }
+
+  /**
+   * Sorts the gradients and hessians according to the sorted order of the
+   * feature.
+   *
+   * @param sortedIndices Vector of indices according to the sorted order of
+   *      the feature.
+   * @param sortedGradients Vector to store the sorted gradients.
+   * @param sortedHessians Vector to store the sorted hessians.
+   */
+  void sortGradAndHess(const arma::uvec& sortedIndices,
+                       arma::vec& sortedGradients,
+                       arma::vec& sortedHessians)
+  {
+    for (size_t i = 0; i < sortedIndices.n_elem; ++i)
+    {
+      sortedGradients[i] = gradients[sortedIndices[i]];
+      sortedHessians[i] = hessians[sortedIndices[i]];
+    }
+  }
+
+  /**
+   * Calculates the sum of hessians and gradients for the left child.
+   *
+   * @return The start index for the iteration to find the best split.
+   */
+  size_t BinaryScanInitialize(const arma::vec& sortedGradients,
+                              const arma::vec& sortedHessians)
+  {
+    size_t index = 0;
+    // Initializing the data members.
+    gLeft = 0;
+    hLeft = 0;
+    while (index < sortedHessians.n_elem - 1 &&
+           hLeft + sortedHessians[index] < minChildWeight)
+    {
+      gLeft += sortedGradients[index];
+      hLeft += sortedHessians[index++];
+    }
+    return index;
+  }
+
+  /**
+   * Steps through the current index and update the statistics. While taking
+   * the step, it ensures that the weight of right child is higher than the
+   * minChildWeight.
+   */
+  void BinaryStep(const arma::vec& sortedGradients,
+                  const arma::vec& sortedHessians,
+                  size_t index,
+                  bool& endLoop)
+  {
+    gLeft += sortedGradients[index];
+    hLeft += sortedHessians[index];
+
+    // If the total weight of right child is lesser than the minChildWeight,
+    // then we need to terminate the loop for iterating over split points.
+    if (hTotal - hLeft < minChildWeight)
+      endLoop = true;
+    return;
+  }
+
+  /**
+   * Calculates the gains for the current split.
+   */
+  double Evaluate()
```

Comment:
Shouldn't this be called `BinaryGains()`?  (I am not 100% sure, I think it should be though.)

Reply:
Yeah, it should be called `BinaryGains()` if you are thinking about making it similar to `MSEGain`, but I thought that we are simply returning the gain of the split so I guessed naming it `Evaluate()` will make more sense. In `MSEGain::BinaryGains()` we were returning gains of both left and right child so we were obligated to use the word "binary" there.

But now I am thinking that we have other functions `BinaryStep()` and `BinaryScanInitialize()` so it would also make sense to name it `BinaryGains()`. I think I will change it to `BinaryGain()`. :)